util/gnmi.go

```diff
@@ -255,3 +255,100 @@ func JoinPaths(prefix, suffix *gpb.Path) (*gpb.Path, error) {
 	}
 	return joined, nil
 }
+
+// CompareRelation describes of the relation between to gNMI paths.
+type CompareRelation int
+
+const (
+	// Equal means the paths are the same.
+	Equal CompareRelation = iota
+	// Disjoint means the paths do not overlap at all.
+	Disjoint
+	// Subset means a path is strictly smaller than the other.
+	Subset
+	// Superset means a path is strictly larger than the other.
+	Superset
+	// PartialIntersect means a path partial overlaps the other.
+	PartialIntersect
+)
+
+// ComparePaths returns the set relation between two paths.
+// It returns an error if the paths are invalid or not one of the relations.
+func ComparePaths(a, b *gpb.Path) CompareRelation {
+	if a.Origin != b.Origin {
+		return Disjoint
+	}
+
+	shortestLen := len(a.Elem)
+
+	// If path a is longer than b, then we start by assuming a subset relationship, and vice versa.
+	var relation CompareRelation
+	if len(a.Elem) > len(b.Elem) {
+		relation = Subset
+		shortestLen = len(b.Elem)
+	} else if len(a.Elem) < len(b.Elem) {
+		relation = Superset
+	}
+
+	for i := 0; i < shortestLen; i++ {
+		elemRelation := comparePathElem(a.Elem[i], b.Elem[i])
+		switch {
+		case elemRelation == PartialIntersect:
+			return PartialIntersect
+		case (relation == Subset && elemRelation == Superset) || (relation == Superset && elemRelation == Subset):
+			return PartialIntersect
+		case elemRelation == Subset || elemRelation == Superset:
+			relation = elemRelation
+		case elemRelation == Disjoint && relation == Equal:
+			relation = Disjoint
```

Comment:
Why is `relation == Equal` a condition? If the current elem is disjoint then the whole path is also disjoint right? Doesn't matter what happened before.

Reply:
simplified even further